src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey.c

```diff
@@ -89,7 +89,220 @@ int32_t CryptoNative_UpRefEvpPkey(EVP_PKEY* pkey)
 }
 
-static bool CheckKey(EVP_PKEY* key, int32_t algId, int32_t (*check_func)(EVP_PKEY_CTX*))
+static bool Lcm(const BIGNUM* num1, const BIGNUM* num2, BN_CTX* ctx, BIGNUM* result)
+{
+    assert(result);
+
+    // lcm(num1, num2) = (num1 * num2) / gcd(num1, num2)
+    BIGNUM* mul = NULL;
+    BIGNUM* gcd = NULL;
+    bool ret = false;
+
+    if ((mul = BN_new()) == NULL ||
+        (gcd = BN_new()) == NULL ||
+        !BN_mul(mul, num1, num2, ctx) ||
+        !BN_gcd(gcd, num1, num2, ctx) ||
+        !BN_div(result, NULL, mul, gcd, ctx))
+    {
+        goto done;
+    }
+
+    ret = true;
+done:
+    BN_clear_free(mul);
+    BN_clear_free(gcd);
+    return ret;
+}
+
+static int32_t QuickRsaCheck(const RSA* rsa, bool isPublic)
+{
+    // This method does some lightweight key consistency checks on an RSA key to make sure all supplied values are
+    // sensible. This is not intended to be a strict key check that verifies a key conforms to any particular set
+    // of criteria or standards.
+
+    const BIGNUM* n = NULL;
+    const BIGNUM* e = NULL;
+    const BIGNUM* d = NULL;
+    const BIGNUM* p = NULL;
+    const BIGNUM* q = NULL;
+    const BIGNUM* dp = NULL;
+    const BIGNUM* dq = NULL;
+    const BIGNUM* inverseQ = NULL;
+    BN_CTX* ctx = NULL;
+
+    // x and y are scratch integers that receive the result of some operations.
+    BIGNUM* x = NULL;
+    BIGNUM* y = NULL;
+
+    // p1 and q1 are to hold p-1 and q-1, respectively. We need these values a couple of times, so don't waste time
+    // recomputing them in scratch integers.
+    BIGNUM* p1 = NULL;
+    BIGNUM* q1 = NULL;
+    int ret = 0;
+
+    RSA_get0_key(rsa, &n, &e, &d);
+
+    // Always need public parameters.
+    if (!n || !e)
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_VALUE_MISSING, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Compatibility: We put this error for OpenSSL 1.0.2 and 1.1.x when the modulus is zero because OpenSSL did not
+    // handle this correctly. Continue to use the same error if the modulus is zero.
+    if (BN_is_zero(n))
+    {
+        ERR_put_error(ERR_LIB_EVP, 0, EVP_R_DECODE_ERROR, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // OpenSSL has kept this value at 16,384 for all versions.
+    if (BN_num_bits(n) > OPENSSL_RSA_MAX_MODULUS_BITS)
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_MODULUS_TOO_LARGE, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Exponent cannot be 1 and must be odd
+    if (BN_is_one(e) || !BN_is_odd(e))
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_BAD_E_VALUE, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // At this point everything that is public has been checked. Mark as successful and clean up.
+    if (isPublic)
+    {
+        ret = 1;
+        goto done;
+    }
+
+    // We do not support validating multi-prime RSA. If there are extra primes (more than two) then treat it as a
+    // decoding failure.
+    if (RSA_get_multi_prime_extra_count(rsa) != 0)
```

Comment:
This may be a slight regression... but I'm OK with that.

If someone was using multi-prime RSA and complains we can circle back and see what we want to do about it.

Reply:
I suppose someone could have a PKCS8 key that has multiple primes in it and it would import successfully. It's easy enough for us to go back to the "old way" in this situation.